private/bufpkg/bufplugin/bufpluginconfig/config.go

```diff
@@ -0,0 +1,191 @@
+		}
+		options[key] = value
+	}
+	runtimeConfig, err := newRuntimeConfig(externalConfig.Runtime)
+	if err != nil {
+		return nil, err
+	}
+	return &Config{
+		Name:    pluginIdentity,
+		Options: options,
+		Runtime: runtimeConfig,
+	}, nil
+}
+
+func newRuntimeConfig(externalRuntimeConfig ExternalRuntimeConfig) (*RuntimeConfig, error) {
+	var (
+		isArchiveEmpty = externalRuntimeConfig.Archive.IsEmpty()
+		isGoEmpty      = externalRuntimeConfig.Go.IsEmpty()
+		isNPMEmpty     = externalRuntimeConfig.NPM.IsEmpty()
+	)
+	if isArchiveEmpty && isGoEmpty && isNPMEmpty {
+		// It's possible that the plugin doesn't have any runtime dependencies.
+		return nil, nil
+	}
+	if isArchiveEmpty && isGoEmpty && !isNPMEmpty {
+		npmRuntimeConfig, err := newNPMRuntimeConfig(externalRuntimeConfig.NPM)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			NPM: npmRuntimeConfig,
+		}, nil
+	}
+	if isArchiveEmpty && !isGoEmpty && isNPMEmpty {
+		goRuntimeConfig, err := newGoRuntimeConfig(externalRuntimeConfig.Go)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			Go: goRuntimeConfig,
+		}, nil
+	}
+	if !isArchiveEmpty && isGoEmpty && isNPMEmpty {
+		archiveRuntimeConfig, err := newArchiveRuntimeConfig(externalRuntimeConfig.Archive)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			Archive: archiveRuntimeConfig,
+		}, nil
+	}
+	// If we made it this far, that means the config specifies multiple
+	// runtime languages.
+	//
+	// We might eventually want to support multiple runtime configuration
+	// (e.g. 'go' and 'archive'), but it's safe to start with an error for
+	// now.
+	return nil, fmt.Errorf("%s configuration contains multiple runtime languages", ExternalConfigFilePath)
+}
+
+func newNPMRuntimeConfig(externalNPMRuntimeConfig ExternalNPMRuntimeConfig) (*NPMRuntimeConfig, error) {
+	if err := validateRuntimeDeps(externalNPMRuntimeConfig.Deps); err != nil {
+		return nil, err
+	}
+	return &NPMRuntimeConfig{
+		Deps: externalNPMRuntimeConfig.Deps,
+	}, nil
+}
+
+func newGoRuntimeConfig(externalGoRuntimeConfig ExternalGoRuntimeConfig) (*GoRuntimeConfig, error) {
+	if err := validateRuntimeDeps(externalGoRuntimeConfig.Deps); err != nil {
+		return nil, err
+	}
+	// The best we can do is verify that the minimum version
+	// is a valid semantic version, just like we do for the
+	// runtime dependencies.
+	//
+	// This will not actually verify that the go version is
+	// in the valid set. It's impossible to capture the
+	// real set of valid identifiers at any given time (for
+	// an old version of the buf CLI) without reaching out to
+	// some external source at runtime.
+	//
+	// Note that this ensures the user's configuration specifies
+	// a 'v' prefix in the version (e.g. v1.18) even though the
+	// minimum version is rendered without it in the go.mod.
+	if externalGoRuntimeConfig.MinVersion != "" && !semver.IsValid(externalGoRuntimeConfig.MinVersion) {
```

Comment:
I think we should loose the `v` and use `golang.org/x/mod/modfile.GoVersionRE.MatchString()` to check the go version.

Reply:
I was hoping that _all_ versions would conform to the same style, as in "all of the runtime dependencies require a `v` prefix, so this should too for consistency". But I felt uneasy about this as I was going about it - I'll adjust this to use the function you just shared and we can go from there.